go/src/koding/klient/machine/transport/rsync/rsync.go

```diff
@@ -0,0 +1,204 @@
+package rsync
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+	"unicode"
+)
+
+// Command describes rsync executable.
+type Command struct {
+	// Download indicates the direction of changes. If set to true, source path
+	// defines remote machine.
+	Download bool
+
+	// SourcePath defines source path from which file(s) will be pulled.
+	// This field is required.
+	SourcePath string
+
+	// DestinationPath defines destination path to which file(s) will be pushed.
+	// This field is required.
+	DestinationPath string
+
+	// Cmd defines command to run. If nil, default rsync command will be used.
+	Cmd *exec.Cmd
+
+	// Username defines remote machine user name. If not set, localhost transfer
+	// will be used.
+	Username string
+
+	// Host defines the remote machine address. If not set, localhost transfer
+	// will be used.
+	Host string
+
+	// PrivateKeyPath if set, SSH remote shell will be used as a data transport.
+	PrivateKeyPath string
+
+	// SSHPort defines custom remote shell port. If not set, default will be used.
+	SSHPort int
+
+	// Progress if set, rsync will be run in recursive and verbose mode. The
+	// current status of downloading will be periodically sent to provided
+	// progress callback function. io.EOF error is sent to the callback when
+	// downloading is complete.
+	Progress func(n, size, speed int64, err error)
+}
+
+// valid checks if command fields are valid.
+func (c *Command) valid() error {
+	if c == nil {
+		return errors.New("rsync: command is nil")
+	}
+
+	if c.SourcePath == "" {
+		return errors.New("rsync: source path is not set")
+	}
+	if c.DestinationPath == "" {
+		return errors.New("rsync: destination path is not set")
+	}
+
+	return nil
+}
+
+// Run starts new rsync process. And waits for it to complete.
+func (c *Command) Run(ctx context.Context) error {
+	if err := c.valid(); err != nil {
+		return err
+	}
+
+	if c.Cmd == nil {
+		c.Cmd = exec.CommandContext(ctx, "rsync")
+		c.Cmd.Env = os.Environ()
+	}
+
+	// Add default arguments.
+	c.Cmd.Args = append(c.Cmd.Args, "-zlptgoDd")
+
+	// Use remote shell if SSH private key path is set.
+	if c.PrivateKeyPath != "" {
+		// TODO(ppknap): check if RC4 cipher will work on every machine without
+		// altering sshd_config on destination.
+		rsh := []string{
+			"ssh", "-T", "-x", "-i", c.PrivateKeyPath,
```

Comment:
If you manage to inject `crypto/ssh` to rsync executable, then it will be possible:). What is wrong with `TODO`? It is about injecting `-c` `arcfour` parameters.